src/relayclient/DevRelayClient.ts

```diff
@@ -0,0 +1,254 @@
+    port: listenPort,
+    backend: relayServer
+  })
+  httpServer.start()
+  return {
+    httpServer,
+    relayServer
+  }
+}
+
+class DevKnownRelays { // extends  KnownRelaysManager {
+  // noinspection JSUnusedGlobalSymbols
+  getRelaysSorted (): RelayRegisteredEventInfo[] {
+    // @ts-ignore
+    return [this.devRelay]
+  }
+
+  // noinspection JSUnusedGlobalSymbols
+  refresh (): void { // ts-ignore
+  }
+
+  devRelay?: RelayRegisteredEventInfo
+}
+
+export interface DevGSNConfig extends Partial<GSNConfig> {
+  relayOwner: Address
+  relayWorkdir?: string
+  relayListenPort?: number
+  relayUrl?: string // defaults to http://localhost:{relayListenPort}
+
+  // TODO: this is actually relay config
+  baseRelayFee?: number
+  pctRelayFee?: number
+  gasPriceFactor: number
+  devMode?: boolean
+}
+
+export class DevRelayClient extends RelayClient {
+  serverStarted: boolean = false
+  httpServer?: HttpServer
+  relayServer?: RelayServer
+  private readonly devConfig: DevGSNConfig;
+
+  constructor (
+    provider: HttpProvider,
+    devConfig: Partial<DevGSNConfig>,
+    overrideDependencies?: Partial<GSNDependencies>
+  ) {
+    super(provider, devConfig,
+      {
+        ...overrideDependencies,
+        knownRelaysManager: new DevKnownRelays() as KnownRelaysManager
+      })
+    this.devConfig = this.config as DevGSNConfig
+  }
+
+  /**
+     * Options include standard transaction params: from,to, gas_price, gas_limit
+     * relay-specific params:
+     *  pctRelayFee (override config.pctRelayFee)
+     *  validateCanRelay - client calls canRelay before calling the relay the first time (defaults to true)
+     *  paymaster - the contract that is compensating the relay for the gas (defaults to transaction destination 'to')
+     * can also override default relayUrl, relayFee
+     * return value is the same as from sendTransaction
+     */
+  async relayTransaction (gsnTransactionDetails: GsnTransactionDetails): Promise<RelayingResult> {
+    await this._initializeRelay()
+    return super.relayTransaction(gsnTransactionDetails)
+  }
+
+  // stop background relay
+  async stopRelay (): Promise<void> {
+    if (!this.serverStarted) {
+      return
+    }
+    if (this.httpServer !== undefined) {
+      this.httpServer.stop()
+      this.httpServer = undefined
+    }
+    if (this.relayServer !== undefined) {
+      // @ts-ignore
+      await this.relayServer.txStoreManager.clearAll()
+      this.relayServer.stop()
+      this.relayServer = undefined
+    }
+  }
+
+  /**
+     * initialize a local relay
+     * @private
+     */
+  async _initializeRelay (): Promise<void> {
+    if (this.serverStarted) {
+      return
+    }
+
+    const hub = await this.contractInteractor._createRelayHub(this.config.relayHubAddress)
+    const stakeManagerAddress = await hub.getStakeManager()
+
+    // flag early, so only the first call will try to bring up a relay
+    // (TODO: other calls should still wait for the relay to start)
```

Comment:
I don't think there is a way to have 'other calls' unless someone creates multiple promises of `relayTransaction` without awaiting on them, which is something we generally never do.